src/tools/wasm-merge.cpp

```diff
@@ -0,0 +1,527 @@
+    // of imports to exports.
+    exportModuleMap[copy.get()] = ExportInfo{inputName, curr->name};
+
+    // Add the export.
+    merged.addExport(std::move(copy));
+  }
+
+  // Start functions must be merged.
+  if (input.start.is()) {
+    if (!merged.start.is()) {
+      // No previous start; just refer to the new one.
+      merged.start = input.start;
+    } else {
+      // Merge them, keeping the order. Note that we need to create a new
+      // function as there may be other references.
+      Builder builder(merged);
+      auto mergedName = Names::getValidFunctionName(merged, "merged.start");
+      auto* oldStart = merged.getFunction(merged.start);
+      auto* oldStartBody = ExpressionManipulator::copy(oldStart->body, merged);
+      auto* newStart = merged.getFunction(input.start);
+      auto* newStartBody = ExpressionManipulator::copy(newStart->body, merged);
+      auto* mergedBody = builder.makeSequence(oldStartBody, newStartBody);
+      auto mergedFunc = builder.makeFunction(
+        mergedName, Signature{Type::none, Type::none}, {}, mergedBody);
+      merged.addFunction(std::move(mergedFunc));
+      merged.start = mergedName;
+    }
+  }
+
+  // TODO: type names, features, debug info, custom sections, dylink info, etc.
+}
+
+// Find pairs of matching imports and exports, and make uses of the import refer
+// to the exported item (which has been merged into the module).
+void fuseImportsAndExports() {
+  // First, scan the exports and build a map. We build a map of [module name] to
+  // [export name => internal name]. For example, consider this module:
+  //
+  //  (module ;; linked in as "module_A"
+  //    (func $foo (export "bar"))
+  //  )
+  //
+  // Then the ModuleExportMap will be:
+  //
+  //  {
+  //    "module_A": {
+  //      "bar": "foo";
+  //    }
+  //  }
+  //
+  using ModuleExportMap = std::unordered_map<Name, NameMap>;
+
+  // A map of ModuleExportMaps, one per item kind (one for functions, one for
+  // globals, etc.).
+  using KindModuleExportMaps =
+    std::unordered_map<ExternalKind, ModuleExportMap>;
+  KindModuleExportMaps kindModuleExportMaps;
+
+  for (auto& ex : merged.exports) {
+    assert(exportModuleMap.count(ex.get()));
+    ExportInfo& exportInfo = exportModuleMap[ex.get()];
+    kindModuleExportMaps[ex->kind][exportInfo.moduleName][exportInfo.baseName] =
+      ex->value;
+  }
+
+  // Find all the imports and see which have corresponding exports, which means
+  // there is an internal item we can refer to. We build up a map of the names
+  // that we should update.
+  KindNameMaps kindNameMaps;
+  ModuleUtils::iterImportable(merged, [&](ExternalKind kind, Importable* curr) {
+    if (curr->imported()) {
+      auto internalName = kindModuleExportMaps[kind][curr->module][curr->base];
+      if (internalName.is()) {
+        // We found something to fuse! Add it to the maps for renaming.
+        kindNameMaps[ModuleItemKind(kind)][curr->name] = internalName;
```

Comment:
Does the conversion from `ExternalKind` to `ModuleItemKind` depend on the enum values being the same? If so, that seems brittle and we might want to make that more explicit with a conversion function or something.

Reply:
Yes, this is an annoyance since C++ has no good way to extend an enum (there's a comment in `wasm.h` on that).

Perhaps an explicit conversion would be a good idea, I agree.

Reply:
If there's at least a comment about keeping them in sync in wasm.h, then it's probably fine.

Reply:
Yes, the comment says they must overlap:

https://github.com/WebAssembly/binaryen/blob/7d5d24f400019ff023b65ccdb3c1d8d07ebb00a5/src/wasm.h#L2071-L2074